ciao-controller/cnci.go

```diff
@@ -166,6 +166,17 @@ func (c *CNCIManager) launch(subnet string) (*types.Instance, error) {
 	return instances[0], nil
 }
 
+// WaitForActiveSubnetString will, given a subnet string, launch a cnci if
+// needed and wait for it to be active, or wait for an existing cnci to become
+// active.
+func (c *CNCIManager) WaitForActiveSubnetString(subnet string) error {
+	subnetInt, err := subnetStringToInt(subnet)
+	if err != nil {
+		return err
+	}
+	return c.WaitForActive(subnetInt)
+}
+
```

Comment:
so this works because the CNCI is being automatically restarted somehow? and we just have to wait for it to come back up?

Reply:
I added this method for restart to ensure that there is a CNCI running before we try to restart an instance.  If no CNCI is running this function will try to launch one.  This would only happen under the following circumstances.

1. The NN is evacuated.
2. The CNCI manager tries to restart the CNCI, it fails as there is no NN.
3. The CNCI is removed.
4. The NN is restored
5. There's nothing that automatically detects this in controller so the CNCI is not automatically restarted until someone starts a new instance.
6. This commit ensures that the CNCI is also restarted when someone restarts an existing instance.

Ideally, when there are no NNs available, the CNCI manager should monitor the network nodes stats and re-launch all the CNCIs it can find for existing tenants, when one comes back online.  But there's nothing that does this right now.

Reply:
There's also some issues related to zombie instances which remain in the controller DB after their node disappears from the cluster, i.e., is not evacuated but just vanishes.  Maybe we need some sort of NodeManager inside controller to handle all of this.

Reply:
probably when we go to implement recovery from node failure we should do this.